docs: add guide on how to build a multi-application dashboard/portal (#3398)
FusionAuth/fusionauth-site: astro/src/content/docs/extend/examples/multi-application-dashboard.mdx

---
title: Multi-Application Dashboard
description: Create a dashboard page with a list of applications a user can log in to.
navcategory: developer
section: extend
subcategory: examples
---

import Aside from 'src/components/Aside.astro';
import Breadcrumb from 'src/components/Breadcrumb.astro';
import IconButton from 'src/components/icon/Icon.astro';
import InlineUIElement from 'src/components/InlineUIElement.astro';

## Overview

If you're using FusionAuth to manage many applications, you might want to provide your users with a central list of applications, so they know what's available. For example, Google does this for its apps.

![Google app selector](/img/docs/extend/examples/multi-application-dashboard/googleGrid.png)

In this guide, you'll learn to make a similar page — a dashboard linking to all the applications in a FusionAuth tenant.

> In the language of authentication, FusionAuth is an identity provider (IdP) and your applications and websites are the service providers (SPs). One way to authenticate a user from a central dashboard is to use an identity-provider-initiated login. In other words, the dashboard will log the user in, and then redirect them to the selected app. In short, a service provider application will receive a login that it didn't initiate itself. This has security complexities and dangers.

Reviewer: ```suggestion
In the language of authentication, FusionAuth is an identity provider (IdP) and your applications and websites are the service providers (SPs). One way to authenticate a user from a central dashboard is to use an IdP-initiated login. In other words, the dashboard will log the user in, and then redirect them to the selected app. In short, a service provider application will receive a login that it didn't initiate itself. This has security implications.
```

> This guide will demonstrate a simpler and safer way to authenticate users from a central dashboard by making each application link in the dashboard point to the application's login page. From that point onward, authentication follows the standard OAuth Authorization Code Flow.

Reviewer: ```suggestion
This guide will demonstrate a simpler and safer way to authenticate users from a central dashboard by making each application link in the dashboard point to the application's login page. From that point onward, authentication follows the standard OAuth Authorization Code grant.
```

> This guide will demonstrate a simpler and safer way to authenticate users from a central dashboard by making each application link in the dashboard point to the application's login page. From that point onwards, authentication follows the standard OAuth Authorization Code Flow.

Reviewer 2: ```suggestion
This guide will demonstrate a simpler and safer way to authenticate users from a central dashboard by making each application link in the dashboard point to the application's login page. From that point onward, authentication follows the standard OAuth Authorization Code Flow.
```

> The next section of this guide will show you how to make a dashboard for two existing FusionAuth applications in the same tenant. All this requires is customizing the theme for the index landing page of the FusionAuth site. The guide will first show you how to install FusionAuth and make two simple web application web servers. If you already have a FusionAuth installation with existing applications, you can skip ahead to the section that creates the dashboard.

Reviewer: ```suggestion
The next section of this guide will show you how to make a dashboard for two existing FusionAuth applications in the same tenant. All this requires is customizing the theme for the index landing page of the FusionAuth site. The guide will first show you how to install FusionAuth and make two simple web applications. If you already have a FusionAuth installation with existing applications, you can skip ahead to [the section that creates the dashboard](#make-a-dashboard).
```

If you want to follow along with the full guide, you will need Docker installed.

## Download The Example Repository And Run FusionAuth

> Use `git clone` to clone the repository at https://github.com/fusionauth/fusionauth-example-multiapp-dashboard, or download and unzip it. 

Reviewer: This repo doesn't exist?

Reviewer: When it does get moved over, we should add it to exampleapps.json as well.

Author: This is the repo https://github.com/ritza-co/fusionauth-example-multiapp-dashboard

We have added it to exampleapps, assuming you keep the same name when moving it to your org

Reviewer: sounds good.This is there now: https://github.com/FusionAuth/fusionauth-example-multiapp-dashboard

Open a terminal in the directory containing the repository files.

Run the command below to start FusionAuth.

>   ```sh

Reviewer: We don't need spaces prepended, do we?

  docker compose up
  ```

<Aside type="note">
  If you have completed any FusionAuth tutorials before, you might need to first delete any existing FusionAuth containers with the same name and the database volumes. Run the command below to do that.

>   ```sh

Reviewer: Same comment about spacing in front of the triple backticks.

  docker ps rm fa faDb; docker compose down -v
  ```
</Aside>

Leave FusionAuth running. 

In a new terminal, run the commands below to start a web server for the Changebank app, which uses FusionAuth for authentication. 

>   ```sh

Reviewer: Same thing with the spaces in front of the backticks.

  cd bankApp

>   docker run --platform=linux/amd64 --rm -v ".:/app" -w "/app" node:23-alpine3.19 sh -c  "npm install"

Reviewer: This won't work on an ARM based chip, I'm guessing. What have you done in the past? can you create a docker compose file for these applications that is cross platform?

Author: it should usually 'just work' through emulation I think (we've tested this on x64 and M1 machines), but yes in general we struggle a bit with figuring out how to do this nicely as we often see little niggles on different platforms with Docker. 

That image is available on all platforms, so maybe we should just add a note at this point advising the reader to match their platform if they have issues?

  docker run --platform=linux/amd64 --rm --network faNetwork -p 3000:3000 -v ".:/app" -w "/app" -e "PORT=3000" node:23-alpine3.19 sh -c  "npm run start"
  ```

> The Changebank app is available at http://localhost:3000. 

Reviewer: ```suggestion
The Changebank app is now running at http://localhost:3000. 
```

> To learn how to make a simple Node.js app that uses FusionAuth, read the [quickstart](/docs/quickstarts/quickstart-javascript-express-web).

Reviewer: ```suggestion
```

Let's not distract the user with a link to the quickstart.

In a third terminal, run the commands below to start a web server for a second app that uses FusionAuth for authentication. The app is called Changeinsurance, and is available at http://localhost:3001.

>   ```sh

Reviewer: Same thing with the spaces in front of the backticks.

  cd insuranceApp

>   docker run --platform=linux/amd64 --rm -v ".:/app" -w "/app" node:23-alpine3.19 sh -c  "npm install"

Reviewer: Same comments about docker run.

  docker run --platform=linux/amd64 --rm --network faNetwork -p 3001:3001 -v ".:/app" -w "/app" -e "PORT=3001" node:23-alpine3.19 sh -c  "npm run start"
  ```

> Before making the dashboard, check that you can log in to all three applications. Either use an incognito browser window or don't enable [Keep me signed in](/docs/lifecycle/authenticate-users/logout-session-management#fusionauth-sso) when logging in, otherwise, you won't see the login form in the rest of this guide:

Reviewer: ```suggestion
Before making the dashboard, check that you can log in to all three applications. Either use an incognito browser window or don't enable <InlineUIElement>[Keep me signed in](/docs/lifecycle/authenticate-users/logout-session-management#fusionauth-sso)</InlineUIElement> when logging in, otherwise, you won't see the login form in the rest of this guide:
```

- Browse to FusionAuth at http://localhost:9011/admin and log in with `admin@example.com` and `password`.
  ![FusionAuth](/img/docs/extend/examples/multi-application-dashboard/fa.png)
- Browse to Changebank at http://localhost:3000 and log in with the same username and password.
  ![Changebank](/img/docs/extend/examples/multi-application-dashboard/changebank.png)
- Browse to Changeinsurance at http://localhost:3001 and log in with the same username and password.
  ![Changeinsurance](/img/docs/extend/examples/multi-application-dashboard/changeinsurance.png)

> If you enabled <InlineUIElement>Keep me signed in</InlineUIElement>, logging out of an application won't log you out of FusionAuth. It will only delete the session of the application. Next time you try to log in, FusionAuth will see the FusionAuth authentication cookie in your browser and automatically log you in.

Reviewer: ```suggestion
If you enabled <InlineUIElement>Keep me signed in</InlineUIElement>, logging out of an application won't necessarily log you out of FusionAuth. It will only delete the session of the application. Next time you try to log in, FusionAuth will see the FusionAuth authentication cookie in your browser and automatically log you in.
```

It will if the application redirects to FusionAuths `/oauth2/logout` URL.

## Make A Dashboard

> Look at the current FusionAuth landing page at http://localhost:9011. 

Reviewer: ```suggestion
Now that you have applications up and running, let's build the dashboard.

Look at the current FusionAuth landing page at http://localhost:9011. 
```

In this section, you will change the FusionAuth landing page to display links to the two banking app web servers you started in the previous section.

> Log in to your [FusionAuth web interface](http://localhost:9011/admin) and browse to <Breadcrumb>Customizations -> Themes</Breadcrumb>. 

Reviewer: thank you for using `<Breadcrumb>`!

> Notice there are three themes in the list. The first two are the default FusionAuth themes. The last one, <InlineUIElement>Bank theme</InlineUIElement>, was added by `kickstart.json` when you ran Docker compose. Read about Kickstart [here](/docs/get-started/download-and-install/development/kickstart).

Reviewer: ```suggestion
Notice there are three themes in the list. The first two are the default FusionAuth themes. The last one, <InlineUIElement>Bank theme</InlineUIElement>, was added by Kickstart when you started FusionAuth up. Read about Kickstart [here](/docs/get-started/download-and-install/development/kickstart).
```

- Click the <IconButton name="edit" /> edit button in the Bank theme row's action column.
- Select the <InlineUIElement>Index</InlineUIElement> page (fourth item in the list on the left).
- Paste the code below into the text box.

>   ```html

Reviewer: This should be remote code pulled from a github repo, not inline FTL.

                  color: white;
                  padding: 10px;
                  margin: 20px;
                  border-radius: 5px;
                  text-align: center;
                  width: 160px;
                  display: inline-block;
                  text-decoration: none;">
                  Changebank
                </a>
                <a href="http://localhost:3001/account" style="background: linear-gradient(to bottom right, skyblue, navy);
                  color: white;
                  padding: 10px;
                  margin: 20px;
                  border-radius: 5px;
                  text-align: center;
                  width: 160px;
                  display: inline-block;
                  text-decoration: none;">
                  Changeinsurance
                </a>
              </div>
          [/@helpers.main]
        [/#if]

      [@helpers.footer]
        [#-- Custom footer code goes here --]
      [/@helpers.footer]
    [/@helpers.body]
  [/@helpers.html]
  ```
- Click the <IconButton name="save" /> save button at the top right.

> Besides a little CSS for styling, the code above has only two changes: Adding hyperlink `<a href="http://localhost:3000/account">Changebank</a>` elements for the two applications. Note that the links don't point to the login page of the target application. Instead, they point to the logged-in main page. If the user isn't logged in to FusionAuth, they will be automatically redirected to the app's login page. If they are already logged in, they will be taken straight to the app. This saves the user time. (To see how this authentication check in the app works, look at `bankApp/routes/index.js` in the sample repository you downloaded.)

Reviewer: ```suggestion
Besides a little CSS for styling, the code above only adds links for the two applications. The links don't point to the login page of the target application. Instead, they point to the logged-in main page. If the user isn't logged in to FusionAuth, they will be automatically redirected to the app's login page. If they are already logged in, they will be taken straight to the app. This saves the user time while still protecting the application content.
```

Reviewer: Please also remote code include `bankApp/routes/index.js` and show how it works, rather than asking the user to look at it.

To see the new landing page, browse to http://localhost:9011.
![FusionAuth app dashboard](/img/docs/extend/examples/multi-application-dashboard/dashboard.png)

> 

Reviewer: ```suggestion

You can extend this to any number of applications, just like Google does.

### Partial Access

You can require application registration for users befeore access is allowed. Do so by enabling the <InlineUIElement>require registration</InlineUIElement> setting in the <Breadcrumb>OAuth</Breadcrumb> tab of the application. 

When this is configured, a user who is not registered to the Changeinsurance application will not be able to access it, even though they can click on the link. They simply won't be logged in.

```

Make sure that the kickstart does set the `require registration` setting for each application.

## Clean Up

To remove all the Docker images used in this guide, stop the running containers and run the commands below.

```sh
docker compose down -v

> docker rmi postgres:16.0-bookworm fusionauth/fusionauth-app:1.51.2 node:23-alpine3.19

Reviewer: I don't think we need to delete the images.
